File: java/src/main/java/jomiel/examples/Options.java
/*
 * -*- coding: utf-8 -*-
 *
 * jomiel-examples
 *
 * Copyright
 *  2019-2021 Toni Gündoğdu
 *
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package jomiel.examples;

import picocli.CommandLine.Command;
import picocli.CommandLine.Option;
import picocli.CommandLine.Parameters;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;

import static java.lang.String.format;
import static java.lang.System.exit;
import static org.tinylog.Logger.info;
import static org.zeromq.ZMQ.getFullVersion;
import static org.zeromq.ZMQ.getVersionString;
import static picocli.CommandLine.usage;

@SuppressWarnings({"CanBeFinal", "FieldMayBeFinal"})
@Command(name = "demo", usageHelpAutoWidth = true)
final class Options implements Callable<Integer> {

    @Option(names = {"-h", "--help"},
            usageHelp = true,
            description = "Display this help message and exit")
    private boolean help = false;

    @Option(names = {"-D", "--print-config"},
            description = "Print configuration values and exit")
    private boolean printConfig = false;

    @Option(names = {"-r", "--router-endpoint"},
            paramLabel = "<addr>",
            description = "Specify the router endpoint address "
                    + "(${DEFAULT-VALUE})")
    String routerEndpoint = "tcp://localhost:5514";

    @Option(names = {"-t", "--connect-timeout"},
            paramLabel = "<timeout>",
            description = "Specify maximum time in seconds for the "
                    + "connection allowed to take "
                    + "(${DEFAULT-VALUE})")
    long connectTimeout = 30;

    @Option(names = {"-V", "--version-zmq"},
            description = "Display ZeroMQ version and exit")
    private boolean zmqVersion = false;

    @Option(names = {"-j", "--output-json"},
            description = "Print dumped messages in JSON")
    boolean outputJson = false;

    @Option(names = {"-c", "--compact-json"},
            description = "Use more compact representation of JSON")
    boolean compactJson = false;

    @Option(names = {"-q", "--be-terse"},
            description = "Be brief and to the point; dump interesting "
                    + "details only")
    boolean beTerse = false;

    @Parameters(arity = "0..*",
            paramLabel = "<uri>",
            description = "the URIs to parse")
    List<String> uri = new ArrayList<>();

    @Override
    public final Integer call() throws Exception {
        if (help) usage(this, System.out);
        else if (printConfig) dumpConfig();
        else if (zmqVersion) printZmqVersion();
        else new Jomiel(this).run();
        return 0;
    }

    private void dumpConfig() throws IllegalAccessException {
        info("---");
        for (final var field : getClass().getDeclaredFields()) {
            field.setAccessible(true);
            info(format("%s: %s",
                    field.getName(),
                    field.get(this)));
        }
        exit(0);
    }

    private void printZmqVersion() {
        info(format("ZeroMQ version %s (%s)",
                getVersionString(),
                getFullVersion()));
        exit(0);
    }
}

File: scala/src/main/scala/jomiel/examples/Options.scala
/*
 * -*- coding: utf-8 -*-
 *
 * jomiel-examples
 *
 * Copyright
 *  2021 Toni Gündoğdu
 *
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package jomiel.examples

import org.tinylog.scala.Logger.info
import org.zeromq.ZMQ.{getFullVersion, getVersionString}
import picocli.CommandLine.{Command, Option, Parameters, usage}

import java.lang.System.out
import java.util.concurrent.Callable

//noinspection VarCouldBeVal
@Command(name = "demo", usageHelpAutoWidth = true)
class Options extends Callable[Int] {

  @Option(names = Array("-h", "--help"),
    description = Array("Display this help message and exit"),
    usageHelp = true
  )
  private var help = false

  @Option(names = Array("-D", "--print-config"),
    description = Array("Print configuration values and exit")
  )
  private var printConfig = false

  @Option(names = Array("-r", "--router-endpoint"),
    paramLabel = "<addr>",
    description =
      Array("Specify the router endpoint address (${DEFAULT-VALUE})")
  )
  var routerEndpoint = "tcp://localhost:5514"

  @Option(names = Array("-t", "--connect-timeout"),
    paramLabel = "<timeout>",
    description =
      Array("Specify maximum time in seconds for the connection "
        + "allowed to take (${DEFAULT-VALUE})")
  )
  var connectTimeout = 30

  @Option(names = Array("-V", "--version-zmq"),
    description = Array("Display ZeroMQ version and exit")
  )
  private var zmqVersion = false

  @Option(names = Array("-j", "--output-json"),
    description = Array("Print dumped messages in JSON")
  )
  var outputJson = false

  @Option(names = Array("-q", "--be-terse"),
    description =
      Array("Be brief and to the point; dump interesting details only")
  )
  var beTerse = false

  @Parameters(arity = "0..*",
    paramLabel = "<uri>",
    description = Array("the URIs to parse")
  )
  var uri = new java.util.ArrayList[String]

  def call(): Int = {
    if (help) {
      usage(this, out)
    } else if (printConfig)
      dumpConfig()
    else if (zmqVersion) {
      printZmqVersion()
    } else {
      new Jomiel(this).run()
    }
    0
  }

  private def dumpConfig(): Unit = {
    info("---")
    for (field <- this.getClass.getDeclaredFields) yield {
      field.setAccessible(true)
      info(s"${field.getName}: ${field.get(this)}")
    }
  }

  private def printZmqVersion(): Unit = {
    info(s"ZeroMQ version $getFullVersion ($getVersionString)")
  }
}
